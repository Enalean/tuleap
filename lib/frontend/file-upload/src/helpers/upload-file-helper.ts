/*
 *  Copyright (c) Enalean, 2024 - Present. All Rights Reserved.
 *
 *  This file is a part of Tuleap.
 *
 *  Tuleap is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  Tuleap is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { type DetailedError, Upload } from "tus-js-client";
import { computedProgress } from "./progress-computation-helper";
import { Option } from "@tuleap/option";
import type { FileUploadOptions, OngoingUpload, OnGoingUploadFile } from "../file-upload-options";

export function uploadFile(
    files: OnGoingUploadFile[],
    file: File,
    upload_href: string,
    onProgressCallback: FileUploadOptions["onProgressCallback"],
    uploaders: Array<Upload>,
): Promise<Option<OngoingUpload>> {
    let uploader: Upload | null;

    const removeUploaderFromList = (): void => {
        if (!uploader) {
            return;
        }

        const index = uploaders.findIndex((file_uploader) => file_uploader === uploader);
        if (index !== -1) {
            uploaders.splice(index, 1);
        }
    };

    return new Promise((resolve, reject): void => {
        uploader = new Upload(file, {
            uploadUrl: upload_href,
            metadata: {
                filename: file.name,
                filetype: file.type,
            },
            onProgress: (bytes_sent: number, bytes_total: number): void => {
                const progress = computedProgress(files, file.name, bytes_sent, bytes_total);
                onProgressCallback(file.name, progress);
            },
            onSuccess: (): void => {
                removeUploaderFromList();
                return resolve(Option.nothing<OngoingUpload>());
            },
            onError: (error: Error | DetailedError): void => {
                return reject(error);
            },
        });
        uploader.start();
        uploaders.push(uploader);
    }).then(() => {
        if (!uploader) {
            throw new Error("The uploader has not been initialized properly.");
        }
        return Option.fromValue({
            cancel: () => {
                uploader?.abort();
                removeUploaderFromList();
            },
        });
    });
}
