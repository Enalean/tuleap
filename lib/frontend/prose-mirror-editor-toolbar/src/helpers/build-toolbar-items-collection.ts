/**
 * Copyright (c) Enalean, 2025 - Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { AdditionalElement, ItemGroup } from "../elements/toolbar-element";
import { ADDITIONAL_ITEMS_GROUP } from "../elements/toolbar-element";

export const buildAdditionalItemGroup = (additional_element: AdditionalElement): ItemGroup => {
    return {
        name: ADDITIONAL_ITEMS_GROUP,
        elements: [additional_element.item_element],
    };
};

export function buildToolbarItems(
    default_item_positons: ItemGroup[],
    additional_elements: AdditionalElement[],
): ItemGroup[] {
    if (additional_elements.length === 0) {
        return default_item_positons;
    }

    const new_item_postions: ItemGroup[] = default_item_positons;
    additional_elements.forEach((additional_item) => {
        const index = new_item_postions.findIndex(
            (item) => item.name === additional_item.target_name,
        );

        if (additional_item.position === "at_the_start") {
            new_item_postions[index].elements.unshift(additional_item.item_element);
        } else if (additional_item.position === "at_the_end") {
            new_item_postions[index].elements.push(additional_item.item_element);
        } else if (additional_item.position === "before") {
            new_item_postions.splice(index, 0, buildAdditionalItemGroup(additional_item));
        } else {
            new_item_postions.splice(index + 1, 0, buildAdditionalItemGroup(additional_item));
        }
    });
    return new_item_postions;
}
