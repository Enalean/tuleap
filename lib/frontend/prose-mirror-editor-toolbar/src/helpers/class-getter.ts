/*
 * Copyright (c) Enalean, 2024 - Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 *
 */

export interface ButtonClasses {
    [key: string]: boolean;
}

export function getClass(is_activated: boolean): ButtonClasses {
    return {
        "tlp-button-primary button-active": is_activated,
        "tlp-button-secondary": !is_activated,
        "prose-mirror-button": true,
        "tlp-button-primary": true,
        "tlp-button-outline": true,
    };
}
