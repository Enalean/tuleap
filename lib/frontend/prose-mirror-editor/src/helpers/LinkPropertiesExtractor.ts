/*
 * Copyright (c) Enalean, 2024 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { DetectLinkNode } from "../plugins/link-popover/helper/LinkNodeDetector";
import type { FindEditorNodeAtPosition } from "../plugins/link-popover/helper/EditorNodeAtPositionFinder";
import type { LinkProperties } from "../types/internal-types";

export type ExtractLinkProperties = {
    extractLinkProperties(position: number): LinkProperties | null;
};

export const LinkPropertiesExtractor = (
    find_editor_node: FindEditorNodeAtPosition,
    detect_link: DetectLinkNode,
): ExtractLinkProperties => ({
    extractLinkProperties: (position): LinkProperties | null => {
        const node = find_editor_node.findNodeAtPosition(position);
        if (!node || !detect_link.isLinkNode(node)) {
            return null;
        }

        const link_mark = node.marks.find((mark) => mark.type.name === "link");
        if (!link_mark) {
            return null;
        }

        return {
            title: node.text ?? "",
            href: link_mark.attrs.href,
        };
    },
});
