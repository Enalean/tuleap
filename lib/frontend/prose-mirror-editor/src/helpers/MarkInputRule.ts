/*
 * Copyright (c) Enalean, 2024 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { InputRule } from "prosemirror-inputrules";
import type { MarkType } from "prosemirror-model";
import type { Transaction } from "prosemirror-state";

type TextPositions = {
    start: number;
    end: number;
};

export const getTextPositions = (match: RegExpMatchArray, start: number): TextPositions => {
    const full_match = match[0];
    const extracted_text = match[1];

    const text_start_position = start + full_match.indexOf(extracted_text);
    const text_end_position = text_start_position + extracted_text.length;

    return {
        start: text_start_position,
        end: text_end_position,
    };
};

export const removeMarkdownCharacters = (
    tr: Transaction,
    start: number,
    end: number,
    text_start_position: number,
    text_end_position: number,
): void => {
    if (text_end_position < end) {
        tr.delete(text_end_position, end);
    }
    if (text_start_position > start) {
        tr.delete(start, text_start_position);
    }
};

export const applyMark = (
    tr: Transaction,
    mark_type: MarkType,
    mark_start_position: number,
    mark_end_position: number,
): void => {
    tr.addMark(mark_start_position, mark_end_position, mark_type.create({}));
    tr.removeStoredMark(mark_type);
};

export function markInputRule(regexp: RegExp, markType: MarkType): InputRule {
    return new InputRule(regexp, (state, match, start, end) => {
        const tr = state.tr;

        const text_positions = getTextPositions(match, start);

        removeMarkdownCharacters(tr, start, end, text_positions.start, text_positions.end);

        const extracted_text = match[1];
        const end_of_mark = start + extracted_text.length;

        applyMark(tr, markType, start, end_of_mark);

        return tr;
    });
}
