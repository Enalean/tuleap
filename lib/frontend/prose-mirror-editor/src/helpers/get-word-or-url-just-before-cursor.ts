/*
 *  Copyright (c) Enalean, 2024 - Present. All Rights Reserved.
 *
 *  This file is a part of Tuleap.
 *
 *  Tuleap is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  Tuleap is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { EditorState } from "prosemirror-state";

export function getWordOrUrlJustBeforeCursor(state: EditorState): string {
    const { $from } = state.selection;
    const node_start = $from.start();
    const cursor_pos = $from.pos;
    const text_from_node_start_to_cursor = state.doc.textBetween(node_start, cursor_pos);
    const text_split_by_space = text_from_node_start_to_cursor.split(" ");
    return text_split_by_space.length > 0
        ? text_split_by_space[text_split_by_space.length - 1]
        : "";
}
