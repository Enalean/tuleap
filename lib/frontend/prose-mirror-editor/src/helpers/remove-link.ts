/*
 * Copyright (c) Enalean, 2024 - Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 *
 */

import { markActive } from "../plugins/toolbar/menu";
import { getWrappingNodeInfo } from "../plugins/toolbar/helper/node-info-retriever";
import type { EditorState, Transaction } from "prosemirror-state";
import { NodeSelection, TextSelection } from "prosemirror-state";
import type { MarkType, ResolvedPos } from "prosemirror-model";

function selectWholeTextLink(transaction: Transaction, from: ResolvedPos): void {
    transaction.setSelection(new NodeSelection(from));
}

function removeMarkOnLink(
    transaction: Transaction,
    from: ResolvedPos,
    to: ResolvedPos,
    mark_type: MarkType,
): void {
    transaction.removeMark(from.pos, to.pos, mark_type);
}

function putBackSelectionOnCursorPosition(
    transaction: Transaction,
    old_from: number,
    old_to: number,
): void {
    transaction.setSelection(
        new TextSelection(transaction.doc.resolve(old_from), transaction.doc.resolve(old_to)),
    );
}

export function removeLink(
    state: EditorState,
    mark_type: MarkType,
    dispatch: (tr: Transaction) => void,
): void {
    if (!markActive(state, mark_type)) {
        return;
    }

    const wrapping_node_info = getWrappingNodeInfo(state.selection.$from, mark_type, state);

    const old_from = state.selection.from;
    const old_to = state.selection.to;

    const from = state.doc.resolve(wrapping_node_info.from);
    const to = state.doc.resolve(wrapping_node_info.to);
    const transaction = state.tr;

    selectWholeTextLink(transaction, from);
    removeMarkOnLink(transaction, from, to, mark_type);
    putBackSelectionOnCursorPosition(transaction, old_from, old_to);

    dispatch(transaction.scrollIntoView());
}
