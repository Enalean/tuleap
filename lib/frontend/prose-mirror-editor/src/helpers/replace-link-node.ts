/*
 * Copyright (c) Enalean, 2024 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { LinkProperties } from "../types/internal-types";
import type { EditorView } from "prosemirror-view";
import { TextSelection } from "prosemirror-state";
import { getWrappingNodeInfo } from "../plugins/toolbar/helper/NodeInfoRetriever";

export function replaceLinkNode(view: EditorView, attrs: LinkProperties): void {
    const schema = view.state.schema;
    const wrapping_node_info = getWrappingNodeInfo(
        view.state.selection.$from,
        schema.marks.link,
        view.state,
    );

    if (!wrapping_node_info.is_creating_node) {
        const from = view.state.doc.resolve(wrapping_node_info.from);
        const to = view.state.doc.resolve(wrapping_node_info.to);
        view.dispatch(view.state.tr.setSelection(new TextSelection(from, to)));
    }

    const node = schema.text(attrs.title, [schema.marks.link.create(attrs)]);
    view.dispatch(view.state.tr.replaceSelectionWith(node, false));
    view.focus();
}
