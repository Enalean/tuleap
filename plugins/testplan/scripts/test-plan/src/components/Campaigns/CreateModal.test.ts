/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { shallowMount } from "@vue/test-utils";
import * as tlp from "tlp";
import type { Modal } from "tlp";
import CreateModal from "./CreateModal.vue";
import type { RootState } from "../../store/type";
import * as tracker_report_retriever from "../../helpers/Campaigns/tracker-reports-retriever";
import { getGlobalTestOptions } from "../../helpers/global-options-for-test";

jest.mock("tlp", () => {
    return {
        __esModule: true,
        createModal: jest.fn(),
    };
});

jest.useFakeTimers();

describe("CreateModal", () => {
    it("Display the modal when mounted", async () => {
        const modal_show = jest.fn();
        jest.spyOn(tlp, "createModal").mockImplementation(() => {
            return {
                show: modal_show,
            } as unknown as Modal;
        });

        const wrapper = shallowMount(CreateModal, {
            global: {
                ...getGlobalTestOptions({
                    state: {
                        milestone_title: "Milestone Title",
                        testdefinition_tracker_id: null,
                    } as RootState,
                }),
            },
        });
        // We need to wait for the loading state to be rendered and the tracker reports status to be resolved
        await jest.runOnlyPendingTimersAsync();

        expect(modal_show).toHaveBeenCalledTimes(1);
        expect(wrapper.element).toMatchSnapshot();
    });

    it("creates the campaign, hides the modal and refresh the backlog items", async () => {
        const modal_hide = jest.fn();
        jest.spyOn(tlp, "createModal").mockImplementation(() => {
            return {
                show: jest.fn(),
                hide: modal_hide,
            } as unknown as Modal;
        });

        const create_campaign_spy = jest.fn();
        const load_backlog_items_spy = jest.fn();

        const wrapper = shallowMount(CreateModal, {
            global: {
                ...getGlobalTestOptions({
                    state: {
                        milestone_title: "Milestone Title",
                        testdefinition_tracker_id: null,
                    } as RootState,
                    modules: {
                        campaign: {
                            namespaced: true,
                            state: {},
                            actions: {
                                createCampaign: create_campaign_spy,
                            },
                        },
                        backlog_item: {
                            namespaced: true,
                            state: {},
                            actions: {
                                loadBacklogItems: load_backlog_items_spy,
                            },
                        },
                    },
                }),
            },
        });

        // We need to wait for the loading state to be rendered and the tracker reports status to be resolved
        await jest.runOnlyPendingTimersAsync();

        wrapper.find("[data-test=new-campaign-label]").setValue("My new campaign");

        await wrapper.trigger("submit");

        expect(create_campaign_spy).toHaveBeenCalledWith(expect.any(Object), {
            label: "My new campaign",
            initial_tests: {
                test_selector: "milestone",
            },
        });
        expect(load_backlog_items_spy).toHaveBeenCalled();
        expect(modal_hide).toHaveBeenCalledTimes(1);
    });

    it("sets an error message when the reports of the test definition tracker cannot be retrieved", async () => {
        const modal_hide = jest.fn();
        jest.spyOn(tlp, "createModal").mockImplementation(() => {
            return {
                show: jest.fn(),
                hide: modal_hide,
            } as unknown as Modal;
        });
        const expected_error = new Error("Something bad happened");
        jest.spyOn(tracker_report_retriever, "getTrackerReports").mockRejectedValueOnce(
            expected_error,
        );

        const wrapper = shallowMount(CreateModal, {
            global: {
                ...getGlobalTestOptions(
                    {
                        state: {
                            milestone_title: "Milestone Title",
                            testdefinition_tracker_id: 12,
                        } as RootState,
                    },
                    (e): void => {
                        expect(e).toBe(expected_error);
                    },
                ),
            },
        });

        // We need to wait for the loading state to be rendered and then to fail to retrieve the tracker reports
        await jest.runOnlyPendingTimersAsync();

        expect(wrapper.find("[data-test=new-campaign-error-message]").exists()).toBe(true);
    });
});
