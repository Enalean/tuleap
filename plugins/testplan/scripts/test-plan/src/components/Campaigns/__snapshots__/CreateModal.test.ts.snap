// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateModal Display the modal when mounted 1`] = `
<form
  aria-labelledby="test-plan-create-campaign-modal-title"
  class="tlp-modal"
  role="dialog"
>
  <div
    class="tlp-modal-header"
  >
    <h1
      class="tlp-modal-title"
      id="test-plan-create-campaign-modal-title"
    >
      Create new campaign
    </h1>
    <button
      aria-label="Close"
      class="tlp-modal-close"
      data-dismiss="modal"
      type="button"
    >
      <i
        aria-hidden="true"
        class="fas fa-times tlp-modal-close-icon"
      />
    </button>
  </div>
  <!--v-if-->
  <div
    class="tlp-modal-body"
  >
    <div>
      <div
        class="tlp-form-element"
      >
        <label
          class="tlp-label"
          for="new-campaign-label"
        >
          Name 
          <i
            aria-hidden="true"
            class="fa-solid fa-asterisk"
          />
        </label>
        <input
          class="tlp-input"
          data-test="new-campaign-label"
          id="new-campaign-label"
          required=""
          type="text"
        />
      </div>
      <create-campaign-test-selector-stub
        initial_tests="[object Object]"
        testdefinition_tracker_reports=""
      />
    </div>
  </div>
  <div
    class="tlp-modal-footer"
  >
    <button
      class="tlp-button-primary tlp-button-outline tlp-modal-action"
      data-dismiss="modal"
      type="button"
    >
      Cancel
    </button>
    <button
      class="tlp-button-primary tlp-modal-action"
      data-test="new-campaign-submit-button"
      type="submit"
    >
      <i
        aria-hidden="true"
        class="fa tlp-button-icon fa-save"
      />
       Create campaign
    </button>
  </div>
</form>
`;
